File: ios/Runner/TuyaLogic/TuyaTimerExtensions.swift
//
//  TuyaTimerExtensions.swift
//  Runner
//
//  Created by Jacob Welin - Wejeo on 02/02/2023.
//

import Foundation

extension TuyaTimerHandler: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        print("Timer Stream started")
        
        self.eventSink = events
//        guard let device = TuyaSmartDevice(deviceId: LocalDataHandler.currentDeviceId ?? "") else{
//            print("Device Error: \(LocalDataHandler.currentDeviceId ?? "")")
//            events(nil)
//            return nil
//        }
        self.getDeviceTimers(completion: { (dataResult) in
            events(dataResult)
        }, failure: { (error) in
            if let e = error?.localizedDescription {
                events(FlutterError.init(code: " tuyaFailureError", message: e, details: nil))
            } else{
                return
            }
        })
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        print("Timer Stream Canceled")
        return nil
    }
    
    
}

File: android/app/src/main/kotlin/dk/wejeo/wejeo_smart/tuya_logic/TuyaTimerExtensions.kt
package dk.wejeo.wejeo_smart.tuya_logic

import android.util.Log
import dk.wejeo.wejeo_smart.TuyaLogic.TuyaTimerHandler
import io.flutter.plugin.common.EventChannel

class TuyaTimerExtensions : EventChannel.StreamHandler, TuyaTimerHandler() {

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        Log.i(LOG_TAG, "Timer Stream started")
        this.eventSink = events
        this.getDeviceTimers { result ->
            if (result == "error") {
                Log.i(LOG_TAG, "RESULT ERROR")

                this.eventSink?.error(
                    "TuyaTimerError",
                    "Could not get timers",
                    "Error when getting timers"
                )
            } else {
                Log.i(LOG_TAG, "RESULT success... $result")

                this.eventSink?.success(result)
            }
        }
    }

    override fun onCancel(arguments: Any?) {
        this.eventSink = null
        Log.i(LOG_TAG, "Home Stream Canceled")
    }
}